analyzers/src/SonarAnalyzer.CSharp/Rules/MessageTemplates/LoggingTemplatePlaceHoldersShouldBeInOrder.cs

```diff
@@ -0,0 +1,114 @@
+            {
+                var templateStart = templateArgument.Expression.GetLocation().SourceSpan.Start;
+                var primaryLocation = Location.Create(context.Tree, new(templateStart + placeholder.Start, placeholder.Length));
+                var secondaryLocation = outOfOrderArgument.GetLocation();
+                context.ReportIssue(Diagnostic.Create(Rule, primaryLocation, [secondaryLocation], placeholder.Name, outOfOrderArgument.ToString()));
+                return; // only raise on the first out-of-order placeholder to make the rule less noisy
+            }
+        }
+    }
+
+    private static IEnumerable<SyntaxNode> PlaceholderValues(InvocationExpressionSyntax invocation, IMethodSymbol methodSymbol)
+    {
+        var placeholderArguments = methodSymbol.Parameters.Where(x => x.Name is "args"
+            || x.Name.StartsWith("argument")
+            || x.Name.StartsWith("propertyValue"));
+        foreach (var placeholderArgument in placeholderArguments)
+        {
+            var paramIndex = methodSymbol.Parameters.IndexOf(placeholderArgument);
+            if (invocation.ArgumentList.Arguments.FirstOrDefault(x => x.NameColon?.GetName() == placeholderArgument.Name) is { } argumentValue)
+            {
+                yield return argumentValue.Expression;
+            }
+            else if (paramIndex < invocation.ArgumentList.Arguments.Count)
+            {
+                var argumentLimit = placeholderArgument.IsParams
+                    ? invocation.ArgumentList.Arguments.Count
+                    : paramIndex + 1;
+                for (var i = paramIndex; i < argumentLimit; i++)
+                {
+                    yield return invocation.ArgumentList.Arguments[i].Expression;
+                }
+            }
+        }
+    }
+
+    private static SyntaxNode OutOfOrderTemplateArgumentForPlaceHolder(MessageTemplatesParser.Placeholder placeholder, int placeholderIndex, ImmutableArray<SyntaxNode> templateArguments)
+    {
+        if (placeholderIndex < templateArguments.Length && PlaceholderMatchesTemplateArgument(placeholder.Name, templateArguments[placeholderIndex]))
+        {
+            return null;
+        }
+        else
+        {
+            for (var i = 0; i < templateArguments.Length; i++)
+            {
+                if (i != placeholderIndex && PlaceholderMatchesTemplateArgument(placeholder.Name, templateArguments[i]))
+                {
+                    return templateArguments[placeholderIndex];
+                }
+            }
+        }
+        return null;
+    }
+
+    private static bool PlaceholderMatchesTemplateArgument(string placeholderName, SyntaxNode templateArgument)
+    {
+        if (templateArgument is MemberAccessExpressionSyntax memberAccess)
+        {
+            return PlaceholderMatchesTemplateArgument(placeholderName, memberAccess.Expression)
+                || PlaceholderMatchesTemplateArgument(placeholderName, memberAccess.Name);
```

Comment:
The `memberAccess.Expression` can be recursive, if it is a `memberAccess` as well.
Please invert the order of the lazy OR:

```csharp
            return PlaceholderMatchesTemplateArgument(placeholderName, memberAccess.Name)
                    || PlaceholderMatchesTemplateArgument(placeholderName, memberAccess.Expression);
```